src/server/pfs/server/obj_block_api_server.go

```diff
@@ -311,22 +387,68 @@ func (s *objBlockAPIServer) ListTags(request *pfsclient.ListTagsRequest, server
 
 	var eg errgroup.Group
 	limiter := limit.New(100)
-	s.objClient.Walk(path.Join(s.localServer.tagDir(), request.Prefix), func(hash string) error {
+	s.objClient.Walk(path.Join(s.localServer.tagDir(), request.Prefix), func(key string) error {
+		tag := filepath.Base(key)
+		if request.IncludeObject {
+			limiter.Acquire()
+			eg.Go(func() error {
+				defer limiter.Release()
+				tagObjectIndex := &pfsclient.ObjectIndex{}
+				if err := s.readProto(key, tagObjectIndex); err != nil {
+					return err
+				}
+				for _, object := range tagObjectIndex.Tags {
+					server.Send(&pfsclient.ListTagsResponse{
+						Tag:    tag,
+						Object: object,
+					})
+				}
+				return nil
+			})
+		}
+		server.Send(&pfsclient.ListTagsResponse{
+			Tag: tag,
+		})
+		return nil
+	})
+	return eg.Wait()
+}
+
+func (s *objBlockAPIServer) DeleteTags(ctx context.Context, request *pfsclient.DeleteTagsRequest) (response *pfsclient.DeleteTagsResponse, retErr error) {
+	func() { s.Log(request, nil, nil, 0) }()
+	defer func(start time.Time) { s.Log(request, response, retErr, time.Since(start)) }(time.Now())
+
+	limiter := limit.New(100)
+	var eg errgroup.Group
+	for _, tag := range request.Tags {
+		tag := tag
 		limiter.Acquire()
 		eg.Go(func() error {
 			defer limiter.Release()
-			tagObjectIndex := &pfsclient.ObjectIndex{}
-			if err := s.readProto(hash, tagObjectIndex); err != nil {
+			tagPath := s.localServer.tagPath(&pfsclient.Tag{tag})
+			if err := s.objClient.Delete(tagPath); err != nil && !s.isNotFoundErr(err) {
 				return err
 			}
-			for _, object := range tagObjectIndex.Tags {
-				server.Send(object)
-			}
 			return nil
 		})
-		return nil
-	})
-	return eg.Wait()
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	return &pfsclient.DeleteTagsResponse{}, nil
+}
+
+func (s *objBlockAPIServer) isNotFoundErr(err error) bool {
+	// GG golang
+	patterns := []string{"not found", "not exist", "NotFound", "NotExist", "404"}
```

Comment:
All implementations of `objectClient.IsNotExist` are fragile anyways... this is sort of like a "catch-all" since it calls `objectClient.IsNotExist` in addition to doing some other checks.  I think there were certain `NotFound` errors that were not caught by the object client, and I thought about just updating the object client, but then I realized that I had no idea if other object clients' `IsNotExist` functions work correctly.  So I just decided to catch the errors at the object server layer just to be safe.  Determining error types across RPC boundaries is just such a pain in the ass.